README.md
# DLOUCFLARE (RECONNSTRUCTION)

To run at a server which has a dynamic IP and cloudflare service. It can dynamic modify your DNS record by your realtime IP address.

## WHAT'S THIS

It's a NODE.JS based project (or even a tool).

Your server is running under a dynamic IP? Want something to modify DNS record to the real-time IP address automatically? Maybe DloucFlare can help you. In fact, ~~**[Zanarberry](http://kacaka.ca/about.html)**~~ is using it.

But first of all, you should have a **CloudFlare** account. We use the first and last letter of `Dynamic` to put on `Cloud`, so we get **DloucFlare**.

## PRE-WORK

Go [http://www.cloudflare.com](http://www.cloudflare.com/) and register for an account. But I think you'd better to know the detail information or features of **CloudFlare** before you use it.

Add your own domain on **CloudFlare** and do as it said.

After you set up your own domain settings on **CloudFlare**, you can use **DloucFlare** now.

## Example

Go to your own **NODE.JS** project directory and get this package from `npm`:

```sh
$ cd /your/path
$ npm install dlouc-flare
```

And then create your `app.js`. The code is very simple:

```javascript
var dloucflare = require("dlouc-flare");
var df = dloucflare.create("your@register.email", "API KEY", "doma.in", "ZONE ID");
df.dynamicDomains([ "sub1", "sub2", "..." ], CHECK_ROUND_TIME);
```

Just start up this `app.js`. If you want it run at the backend, you can use `nohup` command:

```sh
$ nohup node app.js > foo.log 2>&1 &
```

### CUSTOM IP GETTER

You can use your own custom IP getter function. Just set it:

```javascript
df.setIpGetter(function(callback) {
    callback(undefined, "127.0.0.1");
});
```

> You should return IP in the second parameter of callback.
>
> And you should query IP address by your own logic.

We also prepared some default IP Getter function for you. You may use it like:

```javascript
var IPGETTER = dloucflare.IPGETTER;
df.setIpGetter(IPGETTER.TELIZE);
df.setIpGetter(IPGETTER.ALIYUN);
df.setIpGetter(IPGETTER.QQ);
```

We list the prepared getter below:

* TELIZE
* ALIYUN
* SOHU
* IP138
* QQ
* IP5
* WIN7SKY
* SZBENDIBAO
* WHATISMYIP

## CONTACT

If you want contribute to this project, you can fork it!

And if you have some question, you can post it to ISSUES or contact me:

  + Email: admin#xcoder.in
  + Website: http://xcoder.in/

This project is working for my own ~~[ZSG](http://kacaka.ca/). Because it's **China Mobile Broadband**, so maybe **China Telecom Broadband** can't access to it.~~

## OLD VERSION

The v0.0.2 version is [here](https://github.com/XadillaX/dloucflare/tree/0.0.2).

## CONTRIBUTION

You're welcome to make pull request!


History.md

1.0.0 / 2018-07-11
==================

**others**
  * [[`317d7bd`](http://github.com/XadillaX/dloucflare/commit/317d7bdcb9c365213fbd2c92cd05ac6fe3852888)] - docs: update example (XadillaX <<i@2333.moe>>),fatal: No names found, cannot describe anything.



test/test.js
/**
 * Created by XadillaX on 2014/9/18.
 */
var DloucFlare = require("../");

var df = DloucFlare.create("YOUR_EMAIL", "YOUR_TOKEN", "kacaka.ca");
df.dynamicDomains([ "@", "www" ], 1000 * 30, function(err, interval) {
    if(err) {
        return console.log(err);
    }
    console.log("Interval: " + interval);
});


example/test.js
/**
 * Created by XadillaX on 2014/9/18.
 */
var DloucFlare = require("../");

var df = DloucFlare.create("YOUR_EMAIL", "YOUR_TOKEN", "kacaka.ca", "ZONE_ID");
df.dynamicDomains([ "@", "www" ], 1000 * 30, function(err, interval) {
    if(err) {
        return console.log(err);
    }
    console.log("Interval: " + interval);
});


test/ipgetter.js
/**
 * XadillaX created at 2015-09-17 17:58:35 With ♥
 *
 * Copyright (c) 2015 Souche.com, all rights
 * reserved.
 */
require("should");
var spidex = require("spidex");
var IPGETTER = require("../").IPGETTER;

describe("# ip getters", function() {
    this.timeout(0);
    var stdIp;

    before(function(done) {
        spidex.get("http://www.telize.com/ip", {
            charset: "utf8",
            timeout: 120000
        }, function(html) {
            stdIp = html;
            while(stdIp[stdIp.length - 1] === "\n") {
                stdIp = stdIp.substr(0, stdIp.length - 1);
            }

            done();
        }).on("error", function(err) {
            console.error(err);
        });
    });

    for(var key in IPGETTER) {
        if(!IPGETTER.hasOwnProperty(key)) continue;

        (function(key) {
            it(key + " should get ip", function(done) {
                IPGETTER[key](function(err, ip) {
                    (err instanceof Error).should.be.eql(false);
                    ip.should.be.eql(stdIp);
                    done();
                });
            });
        })(key); /* jshint ignore:line */
    }
});


lib/df.js
/**
 * Created with JetBrains WebStorm.
 * User: xadillax
 * Date: 9/7/13
 * Time: 9:35 PM
 * DloucFlare : The dynamic cloudflare domain updater.
 */
var spider = require("nodegrassex");
var querystring = require("querystring");
var logger = require("./logger")

/**
 * The constructor function
 * @param email
 * @param apikey
 * @param domain
 */
function dloucflare(email, apikey, domain) {
    this.email = email;
    this.apikey = apikey;
    this.domain = domain.toLowerCase();

    this.records = null;

    this.autoRunning = false;
};

/**
 * Get the DNS records of one domain.
 * @param callback
 */
dloucflare.prototype.getDNSRecords = function(callback) {
    if(this.autoRunning) {
        if(callback !== undefined) {
            callback.bind(this)(false, "The dloucflare object is in auto running mode", null);
        }
        return;
    }

    /**
     * Get the records via the API of cloudflare.
     * @refer http://www.cloudflare.com/docs/client-api.html#s3.3
     *
     * curl https://www.cloudflare.com/api_json.html \
     *   -d 'a=rec_load_all' \
     *   -d 'tkn=8afbe6dea02407989af4dd4c97bb6e25' \
     *   -d 'email=sample@example.com' \
     *   -d 'z=example.com'
     */
    var param = {
        "a"     : "rec_load_all",
        "tkn"   : this.apikey,
        "email" : this.email,
        "z"     : this.domain
    };
    var paramstr = querystring.stringify(param);

    var url = "https://www.cloudflare.com/api_json.html?" + paramstr;

    var self = this;
    spider.get(url, function(data, status, respheader) {
        /**
         * If the status is wrong.
         */
        if(status !== 200) {
            if(callback !== undefined) {
                callback.bind(self)(false, "Wrong GET status", null);
            }
            return;
        }

        try {
            var json = JSON.parse(data);
        } catch(e) {
            /**
             * If the json string is wrong.
             */
            if(callback !== undefined) {
                callback.bind(self)(false, "Received a wrong JSON string", null);
            }
            return;
        }

        /**
         * An error occurred told by JSON.
         */
        if(json["result"] === "error") {
            var msg = json["msg"];
            if(undefined === msg) msg = "Unknown error";

            if(callback !== undefined) {
                callback.bind(self)(false, msg, null);
            }
            return;
        } else if(json["result"] === "success") {
            var droclass = require("./dro");
            var dro = new droclass(self.email, self.apikey, self.domain);

            /**
             * Push each DNS record to the record array.
             */
            for(var i = 0; i < json["response"]["recs"]["count"]; i++) {
                dro._pushRecord(json["response"]["recs"]["objs"][i]["display_name"], json["response"]["recs"]["objs"][i]);
            }

            if(callback !== undefined) {
                self.records = dro;
                callback.bind(self)(true, "", dro);
            }
        } else {
            /**
             * Unknown error
             * @type {string}
             */
            var msg = "Unknown error";

            if(callback !== undefined) {
                callback.bind(self)(false, msg, null);
            }
            return;
        }
    }, {}, "utf8").on("error", function(e) {
        if(callback !== undefined) {
            callback.bind(self)(false, e.message, null);
        }
        return;
    });
};

/**
 * Modify a record's IP
 * @param name
 * @param ip
 * @param callback
 */
dloucflare.prototype.modifyIP = function(name, ip, callback) {
    /**
     * No record yet.
     */
    if(this.records === null) {
        if(undefined !== callback) {
            callback.bind(this)(false, "No record object yet");
        }
        return;
    }

    /**
     * No such record or not A record
     */
    if(name === this.domain) name = "@";
    var record = this.records.getRecord(name);
    if(record === null) {
        if(undefined !== callback) {
            callback.bind(this)(false, "No such record");
        }
        return;
    }
    if(record.getType() !== "A") {
        if(undefined !== callback) {
            callback.bind(this)(false, "Not an A record");
        }
        return;
    }
    if(name === "@") name = this.domain;

    /**
     * Modify one record
     * @refer
     */
    var param = {
        "a"     : "rec_edit",
        "tkn"   : this.apikey,
        "email" : this.email,
        "id"    : record.getID(),

        "z"     : this.domain,
        "type"  : "A",
        "name"  : name,
        "content" : ip,

        "service_mode" : record["service_mode"],
        "ttl"   : record["ttl"]
    };

    var paramstr = querystring.stringify(param);

    var url = "https://www.cloudflare.com/api_json.html?" + paramstr;

    var self = this;
    spider.get(url, function(data, status, respheader) {
        /**
         * If the status is wrong.
         */
        if(status !== 200) {
            if(callback !== undefined) {
                callback.bind(self)(false, "Wrong GET status");
            }
            return;
        }

        try {
            var json = JSON.parse(data);
        } catch(e) {
            /**
             * If the json string is wrong.
             */
            if(callback !== undefined) {
                callback.bind(self)(false, "Received a wrong JSON string");
            }
            return;
        }

        if(json["result"] === "success") {
            if(name === self.domain) name = "@";
            self.records.obj[name]["content"] = ip;

            if(callback !== undefined) {
                callback.bind(self)(true, "");
            }
            return;
        } else {
            var msg = json["msg"];
            if(msg === undefined || msg === null) {
                msg = "Unknown error";
            }

            if(callback !== undefined) {
                callback.bind(self)(false, msg);
            }
            return;
        }
    }, {}, "utf8").on("error", function(e) {
        if(callback !== undefined) {
            callback.bind(self)(false, e.message);
        }
        return;
    });
};

/**
 * The impl function
 * @param df
 * @param names
 * @param time
 * @private
 */
function _dynamic(df, names, time) {
    /**
     * Stopped.
     */
    if(df.autoRunning !== true) {
        logger.log("The dloucflare object is stopped.", df.domain);
        return;
    }

    /**
     * Get current IP.
     *
     * @refer http://ip-api.com/docs/api:json
     */
    var ipurl = "http://ip-api.com/json";
    spider.get(ipurl, function(data, status, respheader) {
        if(status !== 200) {
            logger.log("Error while getting IP: Wrong status. Skip this round.", df.domain, "WARN");
            setTimeout(_dynamic, time, df, names, time);
            return;
        }

        /**
         * Parse the JSON
         */
        try {
            var json = JSON.parse(data);
        } catch(e) {
            logger.log("Error while getting IP: Wrong JSON data. Skip this round.", df.domain, "WARN");
            setTimeout(_dynamic, time, df, names, time);
            return;
        }

        var ip = json["query"];
        //ip = "1.1.1.1";     ///< test.
        if(json["query"] === undefined) {
            logger.log("Error while getting IP: Wrong JSON data. Skip this round.", df.domain, "WARN");
            setTimeout(_dynamic, time, df, names, time);
            return;
        } else {
            /**
             * Visit all the names in the array.
             */
            for(var i = 0; i < names.length; i++) {
                var nm = names[i];

                /**
                 * If the IP of that record doesn't meet current one, we should modify it.
                 */
                var record = df.records.getRecord(nm);
                if(record.getType() !== "A" || record.getIP() !== ip) {
                    var str = "Subdomain " + nm + " should be modified to " + ip + " (current: ";
                    if(record.getType() !== "A") str += "type " + record.getType() + ").";
                    else str += record.getIP() + ").";

                    logger.log(str, df.domain);

                    df.modifyIP(nm, ip, function(status, msg) {
                        if(!status) {
                            logger.log("Error while modifying subdomain " + nm + ": " + msg + ".", df.domain, "WARN");
                            return;
                        } else {
                            logger.log("Successfully modifying subdomain " + nm, df.domain);
                            return;
                        }
                    });
                }
            }

            setTimeout(_dynamic, time, df, names, time);
        }
    }, {}, "utf8").on("error", function(e) {
        logger.log("Error while getting IP: " + e.message + ". Skip this round.", df.domain, "WARN");
        setTimeout(_dynamic, time, df, names, time);
        return;
    });
}

/**
 * Auto changing domains with their dynamic IP.
 * @param namearray
 */
dloucflare.prototype.dynamicDomains = function(nameArray, timeOut) {
    if(this.autoRunning) {
        logger.log("This dloucflare object is already running.", this.domain, "WARN");
        return;
    }

    this.getDNSRecords(function(status, msg, records) {
        if(!status) {
            logger.log("Error while starting dynamic service: " + msg + ".", this.domain, "ERROR");
            return;
        }

        this.autoRunning = true;

        /**
         * Go to the impl function.
         */
        setTimeout(_dynamic, 1, this, nameArray, timeOut);
    });
}

/**
 * Export object.
 * @type {Function}
 */
module.exports = dloucflare;


lib/dro.js
/**
 * Created with JetBrains WebStorm.
 * User: xadillax
 * Date: 9/7/13
 * Time: 9:35 PM
 * DloucFlare : The dynamic cloudflare domain updater.
 */
function dnsRecordObject(email, apikey, domain) {
    this.obj = [];
    this.email = email;
    this.apikey = apikey;
    this.domain = domain;
};

/**
 * Set the dns records. Used by dlouc flare.
 * @param obj
 * @private
 */
dnsRecordObject.prototype._setRecords = function(obj) {
    this.obj = obj;
}

/**
 * Push one record to the array. Used by dlouc flare.
 * @param obj
 * @private
 */
dnsRecordObject.prototype._pushRecord = function(name, obj) {
    if(name === this.domain) name = "@";
    this.obj[name] = obj;

    /**
     * Add some functions.
     *
     *  + getName
     *  + getWholeDomain
     *  + getType
     *  + getIP
     */
    this.obj[name].getName = function() {
        return name;
    }
    this.obj[name].getWholeDomain = function() {
        return this["display_name"];
    }
    this.obj[name].getType = function() {
        return this["type"];
    }
    this.obj[name].getIP = function() {
        if(this.getType() !== "A") {
            throw "You can't get the IP while this record is not A type.";
        } else {
            return this["content"];
        }
    }
    this.obj[name].getID = function() {
        return this["rec_id"];
    }
}

/**
 * Get a record from array.
 * @param name
 */
dnsRecordObject.prototype.getRecord = function(name) {
    if(this.obj[name] === undefined) return null;
    return this.obj[name];
}

module.exports = dnsRecordObject;


lib/ddns.js
/**
 * Created by XadillaX on 2014/9/18.
 */
var Scarlet = require("scarlet-task");
var IPGETTER = require("./ipgetters");

/**
 * ddns
 * @param dloucflare
 * @constructor
 */
var DDNS = function(dloucflare) {
    this.dloucflare = dloucflare;
    this.getIpFunction = undefined;

    this.scarlet = new Scarlet(10);
};

/**
 * setIpGetter
 * @param {Function} getter the ip getter function
 */
DDNS.prototype.setIpGetter = function(getter) {
    this.getIp = getter;
};

/**
 * get current server ip
 * @param callback
 */
DDNS.prototype.getIp = IPGETTER.ALIYUN;

/**
 * _changeIp
 * @param {TaskObject} TO the scarlet task object
 */
DDNS.prototype._changeIp = function(TO) {
    var task = TO.task;
    var ip = task.ip;
    var dnsRecord = task.object;

    if(dnsRecord.recordIp() === ip) {
        console.log("Domain " + dnsRecord.object.name + "'s ip equals to current server ip: " + ip + ".");
        return this.scarlet.taskDone(TO);
    }

    var self = this;
    dnsRecord.modifyIp(ip, function(err) {
        if(err) {
            console.log(
                "Domain " + dnsRecord.object.name + "(" + dnsRecord.object.content +
                ") can't change to ip " + ip + ": " + err.message);
        } else {
            console.log("Domain " + dnsRecord.object.name + " changed ip to " + ip + ".");
        }

        self.scarlet.taskDone(TO);
    });
};

/**
 * dynamic domains
 * @param subdomains
 * @param interval
 * @param callback
 */
DDNS.prototype.dynamicDomains = function(subdomains, interval, callback) {
    if(callback === undefined) callback = function() {};
    var self = this;

    // get domains
    this.dloucflare.getDNSRecords(function(err, records) {
        if(err) {
            return callback(err);
        }

        var subdomainsObject = subdomains.map(function(subdomain) {
            for(var i = 0; i < records.length; i++) {
                if(records[i].alias === subdomain.toLowerCase()) {
                    if(records[i].recordType() === "A") {
                        return records[i];
                    } else {
                        return undefined;
                    }
                }
            }

            return undefined;
        }).compact(true);

        var _do = function() {
            // get ip
            var ipFunc = (undefined === self.getIpFunction) ? self.getIp : self.getIpFunction;
            ipFunc(function(err, ip) {
                if(err) {
                    return console.log("Error occurred while fetching curreng ip: " + err.message);
                }

                console.log("Current server IP: " + ip);

                for(var i = 0; i < subdomainsObject.length; i++) {
                    self.scarlet.push({
                        ip: ip,
                        object: subdomainsObject[i]
                    }, self._changeIp.bind(self));
                }
            });
        };

        process.nextTick(_do);
        var intervalId = setInterval(_do, interval);

        return callback(undefined, intervalId);
    });
};

module.exports = DDNS;


lib/dloucflare.js
/**
 * Created by XadillaX on 2014/9/18.
 */
var spidex = require("spidex");
var qs = require("querystring");
var DNSRecordObject = require("./dnsrecordobject");

/**
 * Dlouc Flare
 * @param email
 * @param apiKey
 * @param domain
 * @constructor
 */
var DloucFlare = function(email, apiKey, domain, zoneId) {
    this.email = email;
    this.apiKey = apiKey;
    this.domain = domain.toLowerCase();
    this.zoneId = zoneId;

    this.records = [];
};

/**
 * get dns records
 * @param callback
 */
DloucFlare.prototype.getDNSRecords = function(callback) {
    if(undefined === callback) callback = function(){};

    var self = this;
    var param = {
        a       : "rec_load_all",
        tkn     : this.apiKey,
        email   : this.email,
        z       : this.domain
    };

    var url = "https://api.cloudflare.com/client/v4/zones/" + self.zoneId + "/dns_records";
    url = url + "?" + qs.stringify({
        page: 1,
        "per_page": 200,
        order: "type",
        direction: "asc"
    });
    spidex.get(url, {
        charset: "utf8",
        header: {
            "X-Auth-Key": self.apiKey,
            "X-Auth-Email": self.email
        },
        timeout: 10000
    }, function(html, status) {
        if(status !== 200) {
            return callback(new Error("Error status while fetching DNS records."));
        }

        var json;
        try {
            json = JSON.parse(html);
        } catch(e) {
            return callback(new Error("Error while parsing DNS records: " + e.message));
        }
        if(json.success === false) {
            var msg = json.errors ? JSON.stringify(json.errors) : undefined;
            if(undefined === msg) msg = "Unknown error.";
            return callback(new Error(msg));
        } else if(json.success === true) {
            var count = json["result_info"]["total_count"];
            var objects = json.result;
            for(var i = 0; i < count; i++) {
                self.records.push(new DNSRecordObject(self, objects[i]));
            }

            callback(undefined, self.records);
        } else {
            callback(new Error("Unknown error."));
        }
    }).on("error", callback);
};

module.exports = DloucFlare;


lib/dnsrecordobject.js
/**
 * Created by XadillaX on 2014/9/18.
 */
var config = require("../config");
var spidex = require("spidex");

/**
 * DNS record object
 * @param df
 * @param origObject
 * @constructor
 */
var DNSRecordObject = function(df, origObject) {
    this.dloucflare = df;
    this.object = origObject;

    // some information
    this.recordId = this.object.id;
    this.domain = this.object.zone_name;
    this.name = this.object.name;
    this.alias = this.name.replace("." + this.domain, ""); // todo
    if(this.name === this.domain) this.alias = "@";
};

/**
 * record type
 * @returns {*}
 */
DNSRecordObject.prototype.recordType = function() {
    return this.object.type;
};

/**
 * record id
 * @returns {*}
 */
DNSRecordObject.prototype.recordId = function() {
    return this.recordId;
};

/**
 * record ip
 * @returns {*}
 */
DNSRecordObject.prototype.recordIp = function() {
    return this.object.content;
};

/**
 * modify ip
 * @param ip
 * @param callback
 * @returns {*}
 */
DNSRecordObject.prototype.modifyIp = function(ip, callback) {
    if(undefined === callback) callback = function(){};

    if(this.recordType() !== "A") {
        return callback(new Error("Domain must be an A type."));
    }

    var param = {
        type: "A",
        name: this.name,
        content: ip,
        ttl: this.object.ttl,
        proxied: this.object.proxied
    };

    var self = this;

    var url = "https://api.cloudflare.com/client/v4/zones/" + self.dloucflare.zoneId +
     "/dns_records/" + self.recordId;

    spidex.put(url, {
        data: JSON.stringify(param),
        charset: "utf8",
        timeout: 10000,
        header: {
            "X-Auth-Key": self.dloucflare.apiKey,
            "X-Auth-Email": self.dloucflare.email,
            "content-type": "application/json"
        },
    }, function(html, status) {
        if(status !== 200) {
            return callback(new Error("Error status while editing " + self.name + "."));
        }

        var json;
        try {
            json = JSON.parse(html);
        } catch(e) {
            return callback(new Error("Error while parsing editing result: " + e.message));
        }

        if(json.success === true) {
            self.object.content = ip;
            return callback();
        } else {
            var msg = JSON.stringify(json.errors) || "Unknown error.";
            return callback(new Error(msg));
        }
    }).on("error", callback);
};

module.exports = DNSRecordObject;


lib/ipgetters.js
/**
 * XadillaX created at 2015-09-17 17:41:35 With ♥
 *
 * Copyright (c) 2015 Souche.com, all rights
 * reserved.
 */
var spidex = require("spidex");

/**
 * TELIZE
 * @param {Function} callback the callback function
 */
exports.TELIZE = function(callback) {
    var url = "http://www.telize.com/ip";
    spidex.get(url, {
        charset: "utf8",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/;
        if(!regex.test(html)) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, html);
    }).on("error", callback);
};

/**
 * ALIYUN
 * @param {Function} callback the callback function
 */
exports.ALIYUN = function(callback) {
    var url = "http://ip.aliyun.com/service/getIpInfo.php?ip=myip";
    spidex.get(url, {
        charset: "utf8",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var json;

        try {
            json = JSON.parse(html);
            callback(undefined, json.data.ip);
        } catch(e) {
            return callback(e);
        }
    }).on("error", callback);
};

/**
 * SOHU
 *
 * @param {Function} callback the callback function
 */
exports.SOHU = function(callback) {
    var url = "http://txt.go.sohu.com/ip/soip";
    spidex.get(url, {
        charset: "utf8",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /window.sohu_user_ip="(\d{1,3}\.\d{1,3}\.\d{1,3}.\d{1,3})"/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);
};

/**
 * IP138
 * @param {Function} callback the callback function
 */
exports.IP138 = function(callback) {
    var url = "http://1111.ip138.com/ic.asp";
    spidex.get(url, {
        charset: "gbk",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /\[(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\]/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);
};

/**
 * QQ
 * @param {Function} callback the callback function
 */
exports.QQ = function(callback) {
    var url = "http://ip.qq.com/";
    spidex.get(url, {
        charset: "gbk",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /您当前的IP为：<span class="red">(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})<\/span>/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);

};

/**
 * IP5
 * @param {Function} callback the callback function
 */
exports.IP5 = function(callback) {
    var url = "http://www.ip5.me/";
    spidex.get(url, {
        charset: "gbk",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /<div id="ip_addr" style="color:#191970">(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})<\/div>/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);

};

/**
 * WIN7SKY
 * @param {Function} callback the callback function
 */
exports.WIN7SKY = function(callback) {
    var url = "http://win7sky.com/ip/";
    spidex.get(url, {
        charset: "gbk",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /您的IP地址是：\[<font color=#FF0000>(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})<\/font>\]/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);

};

/**
 * SZBENDIBAO
 * @param {Function} callback the callback function
 */
exports.SZBENDIBAO = function(callback) {
    var url = "http://sz.bendibao.com/ip/ip.asp";
    spidex.get(url, {
        charset: "gbk",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /你的电脑的公网IP地址：(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);
};

/**
 * WHATISMYIP
 * @param {Function} callback the callback function
 */
exports.WHATISMYIP = function(callback) {
    var url = "http://www.whatismyip.com.tw/";
    spidex.get(url, {
        charset: "utf8",
        timeout: 120000
    }, function(html) {
        html = html.trim();
        var regex = /<h2>(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})<\/h2>/;
        var regexResult = regex.exec(html);
        if(regexResult.length < 2) {
            return callback(new Error("Not a normal ip address."));
        }

        callback(undefined, regexResult[1]);
    }).on("error", callback);
};
